File: 3.Stacks and Queues/0.implementation_stack.js
class Node {
    constructor(value){
      this.value = value;
      this.next = null;
    }
  }
  
  class StackLinkedList {
    constructor(){
      this.top = null;
      this.bottom = null;
      this.length = 0;
    }
    peek() {
        return this.top;
    }
    push(value){
        let node = new Node(value);
        if(this.length === 0){
            this.top = node;
            this.bottom = node;
        } else {
            let temp = this.top;
            this.top = node;
            this.top.next = temp;
        }
        this.length++;
        return this;
    }
    pop(){
        if(!this.top){
            return null;
        }
        if(this.top === this.bottom){
            this.bottom = null;
        }
        let temp = this.top;
        this.top = this.top.next;
        this.length--;
        return temp;
    }
    isEmpty () {
        return this.length === 0;
    }
  }
  
  const myStack = new Stack();
    myStack.push('google');
    myStack.push('udemy');
    myStack.push('discord');
    console.log(myStack.pop());
    console.log(myStack.peek());
    console.log(myStack.isEmpty());
    console.log(myStack);
  
class StackArray {
    constructor(){
        this.data = [];
    }

    peek() {
        return this.data[this.data.length - 1];
    }

    push(value){
        this.data.push(value);
        return this;
    }

    pop(){
        this.data.pop();
        return this;
    }

    isEmpty () {
        return this.data.length === 0;
    }
}
  

File: 3.Stacks and Queues/0.implementation_stack.py
class Node:
    def __init__(self, value: int) -> None:
        self.value = value
        self.next = None # type: ignore

class StackLinkedList:
    def __init__(self) -> None:
        self.top = None
        self.bottom = None
        self.length = 0

    def peek(self):
        return self.top
    
    def push(self, value: int):
        node = Node(value)
        if self.length == 0:
            self.top = node
            self.bottom = node
        else:
            node.next = self.top
            self.top = node

        self.length += 1
        return self
    
    def pop(self):
        top = self.top
        if self.length == 0:
            return self
        if self.length == 1:
            self.top = None
            self.bottom = None
        else:
            self.top = self.top.next
        self.length -= 1
        return top
    
    def print_stack(self) -> list:
        array = []
        current = self.top
        while current != None:
            array.append(current.value)
            current = current.next
        return array
    
    def is_empty(self) -> bool:
        return self.length == 0
    
class StackArray:
    def __init__(self) -> None:
        self.data = []
    
    def peek(self):
        return self.data[len(self.data) - 1]
    
    def push(self, value):
        self.data.append(value)
        return self
    
    def pop(self):
        return self.data.pop()
    
    def print_stack(self) -> list:
        return self.data
    
    def is_empty(self) -> bool:
        return len(self.data) == 0